File: skinmate(추천해조)/auth/src/main/java/com/skinmate/auth/exception/GlobalExceptionHandler.java
package com.skinmate.auth.exception;

import com.skinmate.auth.domain.ResponseCode;
import com.skinmate.auth.dto.ApiResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {
    
    // @Valid 필드에 발생한 예외 처리
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiResponse<Void>> handleValidationException(MethodArgumentNotValidException ex) {
        String errorMessage = ex.getBindingResult().getAllErrors().get(0).getDefaultMessage();
        log.error("handleValidationException : {}", errorMessage);
        
        return ResponseEntity
                .status(ResponseCode.INVALID_REQUEST.getHttpStatus())
                .body(ApiResponse.error(ResponseCode.INVALID_REQUEST.getCode(), errorMessage));
    }
    
    // CustomException 처리
    @ExceptionHandler(CustomException.class)
    public ResponseEntity<ApiResponse<Void>> handleCustomException(CustomException e) {
        log.error("CustomException: {} - {}", e.getResponseCode(), e.getMessage());
        
        ApiResponse<Void> response = ApiResponse.error(
            e.getResponseCode().getCode(),
            e.getMessage()
        );
        
        return ResponseEntity
            .status(e.getResponseCode().getHttpStatus())
            .body(response);
    }
    
    // 기타 모든 예외 처리
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiResponse<Void>> handleException(Exception e) {
        log.error("Unexpected exception", e);
        
        ApiResponse<Void> response = ApiResponse.error(
            ResponseCode.INTERNAL_SERVER_ERROR.getCode(),
            ResponseCode.INTERNAL_SERVER_ERROR.getMessage()
        );
        
        return ResponseEntity
            .status(ResponseCode.INTERNAL_SERVER_ERROR.getHttpStatus())
            .body(response);
    }
}

File: Bask-EAT((주)식회사)/Bask-EAT-Service-Backend/src/main/kotlin/com/bask_eat/business_backend/exception/GlobalExceptionHandler.kt
package com.bask_eat.business_backend.exception

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import java.util.*

@RestControllerAdvice
class GlobalExceptionHandler {

  @ExceptionHandler(AuthenticationException::class)
  fun handleAuthenticationException(
    ex: AuthenticationException,
    request: WebRequest
  ): ResponseEntity<ErrorResponse> {
    logger.error("Authentication error", ex)

    val errorResponse = ErrorResponse(
      error = "AUTHENTICATION_ERROR",
      message = ex.message ?: "인증 오류가 발생했습니다",
      timestamp = Date(),
      path = request.getDescription(false)
    )

    return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorResponse)
  }

  @ExceptionHandler(ChatException::class)
  fun handleChatException(
    ex: ChatException,
    request: WebRequest
  ): ResponseEntity<ErrorResponse> {
    logger.error("Chat error", ex)

    val errorResponse = ErrorResponse(
      error = "CHAT_ERROR",
      message = ex.message ?: "채팅 처리 중 오류가 발생했습니다",
      timestamp = Date(),
      path = request.getDescription(false)
    )

    return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse)
  }

  @ExceptionHandler(LlmException::class)
  fun handleLlmException(
    ex: LlmException,
    request: WebRequest
  ): ResponseEntity<ErrorResponse> {
    logger.error("LLM error", ex)

    val errorResponse = ErrorResponse(
      error = "LLM_ERROR",
      message = ex.message ?: "LLM 서비스 오류가 발생했습니다",
      timestamp = Date(),
      path = request.getDescription(false)
    )

    return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE).body(errorResponse)
  }

  @ExceptionHandler(MethodArgumentNotValidException::class)
  fun handleValidationExceptions(
    ex: MethodArgumentNotValidException,
    request: WebRequest
  ): ResponseEntity<ErrorResponse> {
    val errorMessage = ex.bindingResult.fieldErrors
      .map { "${it.field}: ${it.defaultMessage}" }
      .joinToString(", ")

    val errorResponse = ErrorResponse(
      error = "VALIDATION_ERROR",
      message = errorMessage,
      timestamp = Date(),
      path = request.getDescription(false)
    )

    return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse)
  }

  @ExceptionHandler(Exception::class)
  fun handleGeneralException(
    ex: Exception,
    request: WebRequest
  ): ResponseEntity<ErrorResponse> {
    logger.error("Unexpected error", ex)

    val errorResponse = ErrorResponse(
      error = "INTERNAL_SERVER_ERROR",
      message = "서버 내부 오류가 발생했습니다",
      timestamp = Date(),
      path = request.getDescription(false)
    )

    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse)
  }

  companion object {
    private val logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)
  }
}

data class ErrorResponse(
  val error: String,
  val message: String,
  val timestamp: Date,
  val path: String
)
